File: 21560 - Sai Suraj/week 5/399_Evaluate Division.py
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        G = defaultdict(dict)
        for i in range(len(equations)):
            a, b = equations[i][0], equations[i][1]
            G[a][b] = values[i]
            G[b][a] = 1 / values[i]
            
        def bfs(start: int,end: int) -> int:
            q = deque()
            q.append((start,1))
            visited = set()
            while q:
                for _ in range(len(q)):
                    node,res = q.popleft()
                    if node==end:
                        return res
                    visited.add(node)
                    for n in G[node]:
                        if n not in visited:
                            q.append((n,res*G[node][n]))
            return -1.0
        
        res = []
        for a,b in queries:
            if a not in G or b not in G:
                res.append(-1.0)
            else:
                res.append(bfs(a,b))
        
        return res

File: 21561 - Suvam Tamang/Week 5/399_Evaluate Division.java
class Solution {
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries){
        double[] result = new double[queries.size()];
        HashMap< String, HashMap<String, Double>> graph = new HashMap<>();
        
        buildGraph(equations, values, graph);
        for(int i=0; i<queries.size(); i++) {
            result[i] = evaluateExpression(queries.get(i).get(0), queries.get(i).get(1), graph,  new ArrayList<String>());
        }
        return result;
    }
    
    public void buildGraph(List<List<String>> equations, double[] values, HashMap<String, HashMap<String, Double>> graph) {
        
        for(int i=0; i<equations.size(); i++) {
            String var1 = equations.get(i).get(0);
            String var2 = equations.get(i).get(1);
        
            if(!graph.containsKey(var1))
                graph.put(var1, new HashMap<String, Double>());
            if(!graph.containsKey(var2))
                graph.put(var2, new HashMap<String, Double>());
            
            graph.get(var1).put(var2, values[i]);
            graph.get(var2).put(var1, 1/values[i]);
        }
    }
    
    public double evaluateExpression(String num1, String num2, HashMap<String, HashMap<String, Double>> graph, ArrayList<String> visited) {
        if( !graph.containsKey(num1) || !graph.containsKey(num2) || visited.contains(num1))
            return -1d;
        if(num1.equals(num2))
            return 1d;
        visited.add(num1);
        
        for (Map.Entry<String, Double> map : graph.get(num1).entrySet()) {
            if(map.getKey().equals(num2))
                return map.getValue();
            
            double res = evaluateExpression(map.getKey(), num2, graph, visited);
            if( res != -1)
                return res * map.getValue();
        }
        return -1d;
    }
}
